File: ports/doom-android/android/app/src/main/java/com/example/doom/DoomModule.java
package com.example.doom;

import android.util.Log;

import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.Arguments;

/**
 * DoomModule - Módulo React Native para exponer la funcionalidad del juego Doom
 * 
 * Este módulo proporciona métodos JavaScript para interactuar con el motor de juego
 * C a través de JNI.
 */
public class DoomModule extends NativeModule {
    private static final String TAG = "DoomModule";
    private static final String NAME = "DoomModule";

    public DoomModule(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @Override
    public String getName() {
        return NAME;
    }

    /**
     * Inicializa el motor de juego
     */
    @ReactMethod
    public void init(int width, int height, Promise promise) {
        try {
            int result = DoomJNI.init(width, height);
            if (result == 1) {
                Log.i(TAG, "Doom engine initialized successfully");
                promise.resolve(true);
            } else {
                Log.e(TAG, "Failed to initialize Doom engine");
                promise.reject("INIT_ERROR", "Failed to initialize Doom engine");
            }
        } catch (Exception e) {
            Log.e(TAG, "Exception in init: " + e.getMessage());
            promise.reject("INIT_ERROR", e);
        }
    }

    /**
     * Limpia y libera recursos del motor de juego
     */
    @ReactMethod
    public void cleanup(Promise promise) {
        try {
            DoomJNI.cleanup();
            Log.i(TAG, "Doom engine cleaned up");
            promise.resolve(true);
        } catch (Exception e) {
            Log.e(TAG, "Exception in cleanup: " + e.getMessage());
            promise.reject("CLEANUP_ERROR", e);
        }
    }

    /**
     * Actualiza el estado del juego
     */
    @ReactMethod
    public void update(int w, int a, int s, int d,
                       int left, int right,
                       int space, int e, int tab, int r,
                       int mouseDx, Promise promise) {
        try {
            DoomJNI.update(w, a, s, d, left, right, space, e, tab, r, mouseDx);
            promise.resolve(true);
        } catch (Exception e) {
            Log.e(TAG, "Exception in update: " + e.getMessage());
            promise.reject("UPDATE_ERROR", e);
        }
    }

    /**
     * Renderiza un frame del juego
     */
    @ReactMethod
    public void render(Promise promise) {
        try {
            DoomJNI.render();
            promise.resolve(true);
        } catch (Exception e) {
            Log.e(TAG, "Exception in render: " + e.getMessage());
            promise.reject("RENDER_ERROR", e);
        }
    }

    /**
     * Obtiene el estado actual del juego
     */
    @ReactMethod
    public void getGameState(Promise promise) {
        try {
            WritableMap state = Arguments.createMap();
            state.putInt("hp", DoomJNI.getPlayerHP());
            state.putInt("ammo", DoomJNI.getPlayerAmmo());
            state.putInt("score", DoomJNI.getPlayerScore());
            state.putBoolean("gameOver", DoomJNI.isGameOver() == 1);
            state.putInt("level", DoomJNI.getLevel());
            
            promise.resolve(state);
        } catch (Exception e) {
            Log.e(TAG, "Exception in getGameState: " + e.getMessage());
            promise.reject("STATE_ERROR", e);
        }
    }

    /**
     * Reinicia el juego
     */
    @ReactMethod
    public void reset(Promise promise) {
        try {
            DoomJNI.reset();
            Log.i(TAG, "Game reset");
            promise.resolve(true);
        } catch (Exception e) {
            Log.e(TAG, "Exception in reset: " + e.getMessage());
            promise.reject("RESET_ERROR", e);
        }
    }

    /**
     * Establece el viewport
     */
    @ReactMethod
    public void setViewport(int width, int height, Promise promise) {
        try {
            DoomJNI.setViewport(width, height);
            promise.resolve(true);
        } catch (Exception e) {
            Log.e(TAG, "Exception in setViewport: " + e.getMessage());
            promise.reject("VIEWPORT_ERROR", e);
        }
    }
}

File: ports/doom-android/android/app/src/main/kotlin/com/example/doom/DoomModule.kt
package com.example.doom

import android.util.Log
import com.facebook.react.bridge.*

/**
 * DoomModule - Módulo React Native en Kotlin para el motor de juego Doom
 *
 * Proporciona métodos JavaScript para interactuar con el motor de juego C/C++
 * a través de JNI con error handling completo y logging.
 */
class DoomModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    companion object {
        private const val TAG = "DoomModule"
        private const val MODULE_NAME = "DoomModule"
    }

    init {
        Log.i(TAG, "DoomModule initialized")
    }

    override fun getName(): String = MODULE_NAME

    /**
     * Inicializa el motor de juego Doom
     *
     * @param width Ancho de la pantalla
     * @param height Alto de la pantalla
     * @param promise Promise para retornar resultado
     */
    @ReactMethod
    fun init(width: Int, height: Int, promise: Promise) {
        try {
            Log.d(TAG, "Initializing Doom engine with dimensions: $width x $height")

            if (width <= 0 || height <= 0) {
                throw IllegalArgumentException("Invalid dimensions: width=$width, height=$height")
            }

            val result = DoomJNI.init(width, height)
            when (result) {
                1 -> {
                    Log.i(TAG, "Doom engine initialized successfully")
                    promise.resolve(true)
                }
                else -> {
                    val error = "Failed to initialize Doom engine (code: $result)"
                    Log.e(TAG, error)
                    promise.reject("INIT_ERROR", error)
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Exception in init", e)
            promise.reject("INIT_ERROR", e.message, e)
        }
    }

    /**
     * Limpia y libera recursos del motor de juego
     *
     * @param promise Promise para retornar resultado
     */
    @ReactMethod
    fun cleanup(promise: Promise) {
        try {
            Log.d(TAG, "Cleaning up Doom engine")
            DoomJNI.cleanup()
            Log.i(TAG, "Doom engine cleaned up successfully")
            promise.resolve(true)
        } catch (e: Exception) {
            Log.e(TAG, "Exception in cleanup", e)
            promise.reject("CLEANUP_ERROR", e.message, e)
        }
    }

    /**
     * Actualiza el estado del juego basado en entrada del usuario
     *
     * @param w Tecla W presionada (booleano como int)
     * @param a Tecla A presionada
     * @param s Tecla S presionada
     * @param d Tecla D presionada
     * @param left Flecha izquierda presionada
     * @param right Flecha derecha presionada
     * @param space Barra espaciadora presionada
     * @param e Tecla E presionada
     * @param tab Tecla Tab presionada
     * @param r Tecla R presionada
     * @param mouseDx Movimiento del mouse en X
     * @param promise Promise para retornar resultado
     */
    @ReactMethod
    fun update(
        w: Int, a: Int, s: Int, d: Int,
        left: Int, right: Int,
        space: Int, e: Int, tab: Int, r: Int,
        mouseDx: Int,
        promise: Promise
    ) {
        try {
            DoomJNI.update(w, a, s, d, left, right, space, e, tab, r, mouseDx)
            promise.resolve(true)
        } catch (e: Exception) {
            Log.e(TAG, "Exception in update", e)
            promise.reject("UPDATE_ERROR", e.message, e)
        }
    }

    /**
     * Renderiza un frame del juego
     *
     * @param promise Promise para retornar resultado
     */
    @ReactMethod
    fun render(promise: Promise) {
        try {
            DoomJNI.render()
            promise.resolve(true)
        } catch (e: Exception) {
            Log.e(TAG, "Exception in render", e)
            promise.reject("RENDER_ERROR", e.message, e)
        }
    }

    /**
     * Obtiene el estado actual del juego
     *
     * @param promise Promise que retorna WritableMap con el estado
     */
    @ReactMethod
    fun getGameState(promise: Promise) {
        try {
            val state = WritableNativeMap().apply {
                putInt("hp", DoomJNI.getPlayerHP())
                putInt("ammo", DoomJNI.getPlayerAmmo())
                putInt("armor", DoomJNI.getPlayerArmor())
                putInt("score", DoomJNI.getPlayerScore())
                putBoolean("gameOver", DoomJNI.isGameOver() == 1)
                putInt("level", DoomJNI.getLevel())
                putInt("enemiesKilled", DoomJNI.getEnemiesKilled())
            }
            Log.d(TAG, "Game state: hp=${state.getInt("hp")}, ammo=${state.getInt("ammo")}")
            promise.resolve(state)
        } catch (e: Exception) {
            Log.e(TAG, "Exception in getGameState", e)
            promise.reject("GET_STATE_ERROR", e.message, e)
        }
    }

    /**
     * Reinicia el juego
     *
     * @param promise Promise para retornar resultado
     */
    @ReactMethod
    fun reset(promise: Promise) {
        try {
            Log.d(TAG, "Resetting game")
            DoomJNI.reset()
            Log.i(TAG, "Game reset successfully")
            promise.resolve(true)
        } catch (e: Exception) {
            Log.e(TAG, "Exception in reset", e)
            promise.reject("RESET_ERROR", e.message, e)
        }
    }

    /**
     * Establece el viewport (área de renderizado)
     *
     * @param width Ancho del viewport
     * @param height Alto del viewport
     * @param promise Promise para retornar resultado
     */
    @ReactMethod
    fun setViewport(width: Int, height: Int, promise: Promise) {
        try {
            if (width <= 0 || height <= 0) {
                throw IllegalArgumentException("Invalid viewport dimensions: width=$width, height=$height")
            }
            DoomJNI.setViewport(width, height)
            Log.d(TAG, "Viewport set to $width x $height")
            promise.resolve(true)
        } catch (e: Exception) {
            Log.e(TAG, "Exception in setViewport", e)
            promise.reject("VIEWPORT_ERROR", e.message, e)
        }
    }

    /**
     * Obtiene estadísticas de rendimiento
     *
     * @param promise Promise que retorna WritableMap con estadísticas
     */
    @ReactMethod
    fun getStats(promise: Promise) {
        try {
            val stats = WritableNativeMap().apply {
                putDouble("fps", DoomJNI.getFPS())
                putInt("drawCalls", DoomJNI.getDrawCalls())
                putInt("triangles", DoomJNI.getTriangleCount())
            }
            promise.resolve(stats)
        } catch (e: Exception) {
            Log.e(TAG, "Exception in getStats", e)
            promise.reject("STATS_ERROR", e.message, e)
        }
    }

    /**
     * Habilita o deshabilita el debug
     *
     * @param enabled True para habilitar debug
     * @param promise Promise para retornar resultado
     */
    @ReactMethod
    fun setDebug(enabled: Boolean, promise: Promise) {
        try {
            DoomJNI.setDebug(if (enabled) 1 else 0)
            Log.d(TAG, "Debug mode set to $enabled")
            promise.resolve(true)
        } catch (e: Exception) {
            Log.e(TAG, "Exception in setDebug", e)
            promise.reject("DEBUG_ERROR", e.message, e)
        }
    }
}
